### ISSUE TITLE
Can't retrieve a previous versions of deleted files

### ISSUE THREAD

**Issue by JElchison** [JElchison/export-surround-to-git#34]
Let's say that a file was deleted in version 2.  `sscm get` won't let you retrieve version 1, even though the GUI does.  (So, right now, we can detect that a file was deleted, we just can't pull the history of its file contents.)

We need a way to do this.

Is this fixed in newer versions of Surround?

**Comment by wyattearp** [comment]
Are you performing the detection with history at the repository level? The "restore" command might be what you're looking for.  Additionally, and it's highly inefficient, if you know what "time" the repository had the file, you could generate a temporary branch at that point in time (in sscm) and then retrieve the file from there (remember, highly inefficient).

**Comment by JElchison** [comment]
You are correct--I can retrieve v1 of the (above-mentioned) example file if I first undelete it.  In order to avoid changing history in the branch of interest, I could do this un-delete in another (new) branch.  However, it makes me sad to think that the export operation isn't read-only.  As the newly created branch would also appear in Git.  And there would be one subbranch for every deleted file in the entire history of each branch.  Ugh.  The branch could be deleted afterward, but it may not be able to be permanently deleted per server permissions.

So far, my list of workarounds:
1. Un-delete the file in the branch of interest.  Fetch the previous versions.  Rollback the file (so that it's deleted again) afterwards.  (Modifies the branch of interest.  Bad things if export interrupted or failure.)
2. Create a new sub-branch for each deleted file.  Delete the subbranch later.  (Doesn't scale well ... lots of sub-branches.)

Do you have any experience with other versions of sscm?  All of my testing has been with sscm 2013.0.0 Build 23 on Linux.  Certainly there must be a clean way to do this using the command line tool...

**Comment by wyattearp** [comment]
I don't have a functional sscm client up and running, but what about trying to do a get based on time instead of version from the repository? The documentation at their site seems to indicate that this is probably a bug / unsupported feature to request it by version.

**Comment by JElchison** [comment]
Good idea, will investigate.  Reference:  http://help.seapine.com/surround/2014.1.0/scmcli/Default.htm#CLI%20Guide/get.htm%3FTocPath%3D_____44

Another idea:  Do it from the API ... http://help.seapine.com/surround/2014.1.0/scmclient/Default.htm#UserGuide/UsingCAPI.htm#kanchor495

**Comment by JElchison** [comment]
Confirmed, you can retrieve a file with timestamp via `-s"yyyymmddhh:mm:ss" -i`.  This is a path forward, but is still not as accurate as I'd like due to #14.

What "seconds" do we provide?
- If we choose randomly, we have a 50% chance (at best) of being wrong.
- If we choose "59", then we may skip past the next newer event in the file's history (if multiple modifications within the same minute).
  - Also, "59" may not acount for events that occurred in the 1.00s after 59.00, but before 0.00

I think I'll still look in to writing a separate tool that uses the API to retrieve a deleted file based on version.

**Comment by JElchison** [comment]
Drawback to using the API:  We'll have to ask the user for his/her SCM creds, instead of using the cached defaults.

**Comment by wyattearp** [comment]
If I recall, isn't the deletion information stored on the repository, not the file? So if you ask the repository when a file was deleted, it should give you a time (that of course will probably have issues with #14), and again ... stupidly inefficient, you could try to get the file with that version from sscm ... if it's not there, it should fail and then you can adjust your get at time to be time the mm:59 before the time provided. _shudder_

**Comment by JElchison** [comment]
I think you're right, but the wheels fall off if more than one operation occurred in the same minute.  Or, worse, in the same second.  Eek.

It's a path forward, but I think I'll see if the API can give us a nice solution to #14 first...

**Comment by JElchison** [comment]
With the new `sscmhist` C application (which will give us seconds in timestamps), this can be solved via `-s"yyyymmddhh:mm:ss" -i`.

Still need to give some thought to whether we should add/subtract a second to ensure that the file we `get` matches the version of interest.

**Comment by ghost** [comment]
In the latest release, 2017.1.0, the following is mentioned as a bug fix "The Get_Repo_By_Timestamp API call did not return files removed after the specified timestamp.".  It sounds like it might apply to this important roadblock of an issue.

https://www.perforce.com/products/surround-scm/release-notes#tab-panel-5781
